clang/tools/archgen-mlir/ArchGenMLIR.cpp

```diff
@@ -0,0 +1,927 @@
+    llvm_unreachable("unhandled Stmt");
+  }
+};
+
+/// List all Types and Statements that needs to be generate for this to work
+struct ASTLister : clang::StmtVisitor<ASTLister, void> {
+  MLIRGenState &state;
+  llvm::DenseSet<clang::QualType> Types;
+  std::string TypeStr;
+  llvm::raw_string_ostream TypeOS;
+  llvm::DenseSet<int /*Stmt::StmtClass*/> Stmts;
+  std::string StmtStr;
+  llvm::raw_string_ostream StmtOS;
+
+  ASTLister(MLIRGenState &s) : state(s), TypeOS(TypeStr), StmtOS(StmtStr) {}
+
+  void addType(clang::QualType Ty, clang::SourceLocation Loc) {
+    if (Ty->isFunctionPointerType())
+      return;
+    mlir::Type MLIRTy = state.getMLIRType(Ty);
+    if (Types.insert(Ty).second) {
+      Loc.print(TypeOS, state.ASTctx.getSourceManager());
+      TypeOS << "\n";
+      MLIRTy.print(TypeOS);
+      TypeOS << "\n";
+      Ty->dump(TypeOS, state.ASTctx);
+      TypeOS << "\n";
+    }
+  }
+
+  void VitisStmt(clang::DeclRefExpr* DRE) {
+    /// This will add the function to the list of functions to "emit". but in
+    /// this case functions are not emitted but there statement and types are
+    /// listed
+    if (auto* FD = dyn_cast<FunctionDecl>(DRE->getDecl()))
+      state.GetOrCreateFunc(FD);
+
+    VisitStmt(DRE);
+  }
+
+  void VisitStmt(clang::Stmt *S) {
+    if (auto *E = dyn_cast<clang::Expr>(S))
+      addType(E->getType(), E->getExprLoc());
+    if (Stmts.insert(S->getStmtClass()).second) {
+      S->dump(StmtOS, state.ASTctx);
+      StmtOS << "\n";
+    }
+    for (auto *Inner : S->children())
+      VisitStmt(Inner);
+  }
+
+  void print(llvm::raw_ostream &os) {
+    os << "Types(" << Types.size() << "):\n";
+    os << TypeOS.str();
+    os << "\n\nStms(" << Stmts.size() << "):\n";
+    os << StmtOS.str();
+  }
+
+  void dump() { print(llvm::errs()); }
+};
+
+class ArchGenMLIRConsumer : public clang::ASTConsumer {
+  clang::CompilerInstance &CI;
+  std::unique_ptr<clang::ASTConsumer> llvmirAstConsumer;
+  llvm::Module *LLVMModule;
+  llvm::LLVMContext *LLVMCtx;
+  MLIRGenState state;
+  MLIREmitter Visitor;
+  ASTLister Lister;
+
+public:
+  ArchGenMLIRConsumer(clang::CompilerInstance &CompilerInstance,
+                      std::unique_ptr<clang::ASTConsumer> Consumer,
+                      llvm::Module *LLVMModule)
+      : CI(CompilerInstance), llvmirAstConsumer(std::move(Consumer)),
+        LLVMModule(LLVMModule), LLVMCtx(&LLVMModule->getContext()),
+        state(CI, LLVMModule), Visitor(state), Lister(state) {}
+
+  void addToEmit(clang::FunctionDecl *FD) {
+    if (ListNotEmit)
+      Lister.addType(FD->getType(), FD->getLocation());
+    else
+      state.GetOrCreateFunc(FD);
+  }
+
+  void FillFunctionBody(clang::FunctionDecl *FD) {
+    if (ListNotEmit)
+      Lister.VisitStmt(FD->getBody());
+    else
+      Visitor.Emit(FD, state.GetOrCreateFunc(FD));
+  }
+
+  /// This is transformation fixes function parameters. it is not generic, it
+  /// exist only because we do not generate parameter properly.
```

Comment:
```suggestion
  /// This transformation fixes the function parameters. It is not generic, it
  /// exist only because we do not generate parameter properly.
```
Btw, is it explained at the place where the generation is done?

Reply:
now yes

Reply:
At least fix the comment.